### PR TITLE
Add comma for required_one_of

### LINKED ISSUE #39786: kubernetes module broken in Ansible 2.5: one of the following is required: f, i, l, e, _, r, e, f, e, r, e, n, c, e"
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

kubernetes

##### ANSIBLE VERSION

```
ansible 2.5.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, Feb 20 2018, 09:19:12) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### CONFIGURATION

```
ANSIBLE_PIPELINING(/etc/ansible/ansible.cfg) = True
DEFAULT_LOG_PATH(/etc/ansible/ansible.cfg) = /var/log/ansible.log
DEFAULT_VAULT_PASSWORD_FILE(env: ANSIBLE_VAULT_PASSWORD_FILE) = /root/.vault_pass.txt
RETRY_FILES_ENABLED(/etc/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT

Red Hat Enterprise Linux 7.5

##### SUMMARY

Ansible module `kubernetes` is broken in Ansible 2.5 while the new `k8s_raw` module is not yet fully functional (see #39674 and others).

##### STEPS TO REPRODUCE

```yaml
- name: "Deploy Hazelcast"
  delegate_to: localhost
  run_once: yes
  kubernetes:
    state: present
    api_endpoint: '{{ dynenv_k8s_endpoint }}'
    inline_data:
      apiVersion: extensions/v1beta1
      kind: Deployment
      metadata:
        name: hazelcast
        namespace: 'dynenv-{{ dynenv_name }}'
        annotations:
          description: Hazelcast
      spec:
        replicas: 1
        template:
          metadata:
            labels:
              name: hazelcast
              app_version: '{{ dynenv_hazelcast_version |string }}'
            annotations:
              prometheus.io/port: '1234'
              prometheus.io/scrape: 'true'
          spec:
            affinity:
              podAntiAffinity:
                requiredDuringSchedulingIgnoredDuringExecution:
                  - labelSelector:
                      matchExpressions:
                      - key: "name"
                        operator: In
                        values:
                        - hazelcast
                    topologyKey: "kubernetes.io/hostname"
            imagePullSecrets:
            - name: dcr
            initContainers:
            - name: wait-for-etcd
              image: dcr.intern.hse24.de/hse24/rhel7
              command: ['sh', '-c', 'until curl --silent --fail http://etcd:2379/health ; do echo waiting for etcd; sleep 2; done;']
            containers:
            - name: hazelcast
              image: '{{ dynenv_hazelcast_image }}:{{ dynenv_hazelcast_version }}'
              args:
                - -d
                - -p
                - "5701"
                - -e
                - "http://etcd:2379"
                - -j
                - "-javaagent:/usr/java/default/lib/ext/jmx_prometheus_javaagent.jar=1234:/etc/jmx_exporter/default.yml -Xms{{ dynenv_hazelcast_heap_size_mb }}m -Xmx{{ dynenv_hazelcast_heap_size_mb }}m -XX:+AlwaysPreTouch -Dhazelcast.http.healthcheck.enabled=true"
              imagePullPolicy: Always
              env:
              - name: LANG
                value: en_US.UTF-8
              - name: TZ
                value: '{{ dynenv_timezone }}'
              resources:
                requests:
                  memory: "{{ dynenv_hazelcast_heap_size_mb }}Mi"
                  # CPU is in "milli-cpu"
                  cpu: "100m"
              ports:
                - containerPort: 80
            restartPolicy: Always
            readinessProbe:
              failureThreshold: 5
              httpGet:
                path: /hazelcast/health
                port: 5701
                scheme: HTTP
              initialDelaySeconds: 15
              timeoutSeconds: 5
            livenessProbe:
              failureThreshold: 5
              httpGet:
                path: /hazelcast/health
                port: 5701
                scheme: HTTP
              initialDelaySeconds: 15
              timeoutSeconds: 5
```

##### EXPECTED RESULTS

Although deprecated, the `kubernetes` module should continue to work as in Ansible 2.4.

##### ACTUAL RESULTS

Playbook failed:

```
TASK [dynenv-hazelcast : Setup Kubernetes service for Hazelcast] **********************************************************************************************************************************************
fatal: [server -> localhost]: FAILED! => {"changed": false, "msg": "one of the following is required: f, i, l, e, _, r, e, f, e, r, e, n, c, e"}
```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Fixes: #39786

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/clustering/k8s/_kubernetes.py

##### ANSIBLE VERSION

```
2.6devel
```

### CODE REVIEW

**Comment by dagwieers** [comment]
IMO if this would fix that issue, it should be fixed in Ansible and not in the module itself.

**Review by dagwieers — CHANGES_REQUESTED** [review]
Should be fixed in Ansible IMO, not in this module.

File: lib/ansible/modules/clustering/k8s/_kubernetes.py [hunk 1]
```diff
@@ -349,7 +349,7 @@ def main():
         mutually_exclusive=(('file_reference', 'inline_data'),
                             ('url_username', 'insecure'),
                             ('url_password', 'insecure')),
-        required_one_of=(('file_reference', 'inline_data')),
+        required_one_of=(('file_reference', 'inline_data'),),
```
Comment (webknjaz) [inline lib/ansible/modules/clustering/k8s/_kubernetes.py:352]: I'd add a space after comma to be compliant with PEP8.
But for readability I'd also make it multiline:
```python
        required_one_of=(
            ('file_reference', 'inline_data'),
        ),
```

Comment (dagwieers) [reply lib/ansible/modules/clustering/k8s/_kubernetes.py:352]: Still I think this is the wrong fix and we should not be adding the comma here.
Ansible should simply handle this case correctly.

Comment (webknjaz) [reply lib/ansible/modules/clustering/k8s/_kubernetes.py:352]: It adds ambiguity. It looks like you're trying to break some PEP 20 rules:
- https://zen-of-python.info/explicit-is-better-than-implicit.html#2
- https://zen-of-python.info/special-cases-arent-special-enough-to-break-the-rules.html#8
- https://zen-of-python.info/in-the-face-of-ambiguity-refuse-the-temptation-to-guess.html#12
- https://zen-of-python.info/there-should-be-one-and-preferably-only-one-obvious-way-to-do-it.html#13

Comment (dagwieers) [reply lib/ansible/modules/clustering/k8s/_kubernetes.py:352]: Welcome in the real world, where rules of thumb are rules of thumb and not religious dogmas.

Here is another one: Robustness principle (Postel's Law): *Be liberal in what you accept, and strict in what you send.*
https://en.wikipedia.org/wiki/Robustness_principle

Comment (webknjaz) [reply lib/ansible/modules/clustering/k8s/_kubernetes.py:352]: Yes, there's Criticism section there.

I'm not in favor encouraging people to send random stuff into functions, which leads to misleading understanding of what it does to the code reader and created way more diversity over codebase.

Anyway @Akasurde has already suggested a good validation solution to this one.

Comment (dagwieers) [reply lib/ansible/modules/clustering/k8s/_kubernetes.py:352]: We're not sending random stuff into functions. We are making the interface (which is codified) accept more than just lists. Which is something we have done in multiple places already.

Comment (dagwieers) [reply lib/ansible/modules/clustering/k8s/_kubernetes.py:352]: Where did @Akasurde suggest an alternative ?

Comment (webknjaz) [reply lib/ansible/modules/clustering/k8s/_kubernetes.py:352]: API-wise it's a terrible idea to accept multiple argument types.

He suggested this in Slack and it's related to #39924. I think @Akasurde will update the PR once he has time.

Comment (dagwieers) [reply lib/ansible/modules/clustering/k8s/_kubernetes.py:352]: @webknjaz In life, it's a terrible idea to speak in absolutes. One day you'll figure that out...

In the meantime you can rewrite Ansible. I'd start with modifying everything in Ansible that accepts single types as a lists. That is *EVERY* YAML list as input. Good luck with that...

**Comment by fabianvf** [comment]
Looks good to me :shipit:

**Review by webknjaz — APPROVED** [review]

**Comment by jiaqingw** [comment]
when can this be merged to main ansible release?

**Comment by webknjaz** [comment]
I think we can merge this right away.

**Comment by mattclay** [comment]
@sivel How difficult would it be to check this in the module validator?

**Comment by sivel** [comment]
> @sivel How difficult would it be to check this in the module validator?

Not extremely.  We already have access to the args and kwargs passed to `AnsibleModule`, we just need to expand the validation:

```
spec, args, kwargs = get_argument_spec(self.path)
```

Assuming someone implements these checks, we should likely enforce that `AnsibleModule` is also called with kwargs.

**Comment by sivel** [comment]
I'll play around with it and open a new PR.

**Comment by webknjaz** [comment]
FTR this has been backported to 2.6

**Pull request merged**